orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RollingRedBlackStrategy.groovy

```diff
@@ -1,4 +1,136 @@
+import com.netflix.spinnaker.orca.pipeline.model.Execution
+import com.netflix.spinnaker.orca.pipeline.model.Stage
+import com.netflix.spinnaker.orca.pipeline.model.SyntheticStageOwner
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.context.ApplicationContext
+import org.springframework.context.ApplicationContextAware
+import org.springframework.stereotype.Component
+
+import static com.netflix.spinnaker.orca.pipeline.StageDefinitionBuilder.StageDefinitionBuilderSupport.newStage
+
+@Slf4j
+@Component
+class RollingRedBlackStrategy implements Strategy, ApplicationContextAware {
+  final String name = "rollingredblack"
+
+  @Autowired
+  DisableClusterStage disableClusterStage
+
+  @Autowired
+  ResizeServerGroupStage resizeServerGroupStage
+
+  @Autowired
+  DetermineTargetServerGroupStage determineTargetServerGroupStage
+
+  @Autowired
+  SourceResolver sourceResolver
+
+  @Override
+  <T extends Execution<T>> List<Stage<T>> composeFlow(Stage<T> stage) {
+    def stages = []
+    def stageData = stage.mapTo(RollingRedBlackStageData)
+    def cleanupConfig = AbstractDeployStrategyStage.CleanupConfig.fromStage(stage)
+
+    Map baseContext = [
+        (cleanupConfig.location.singularType()): cleanupConfig.location.value,
+        cluster                                : cleanupConfig.cluster,
+        credentials                            : cleanupConfig.account,
+        cloudProvider                          : cleanupConfig.cloudProvider,
+    ]
+
+    Map savedCapacity = stageData.savedCapacity ?: stageData.capacity
+    if (!savedCapacity) {
+      savedCapacity = [
+          min: stageData.targetSize,
+          max: stageData.targetSize,
+          desired: stageData.targetSize
+      ]
+    }
+
+    if (stageData.targetSize) {
+      stage.context.targetSize = 0
+    }
+
+    stage.context.savedCapacity = savedCapacity
+    stage.context.capacity = [
+        min: 0,
+        max: 0,
+        desired: 0
+    ]
+
+    def targetPercentages = stageData.getTargetPercentages()
+    if (targetPercentages.size() == 0 || targetPercentages[-1] != 100) {
+      log.info("Inserting implicit 100% final target percentage...")
+      targetPercentages.add(100)
+    }
+
+    def findContext = baseContext + [
+        target: TargetServerGroup.Params.Target.current_asg_dynamic,
+        targetLocation: cleanupConfig.location,
+    ]
+
+    stages << newStage(
+        stage.execution,
+        determineTargetServerGroupStage.type,
+        "Determine Deployed Server Group",
+        findContext,
+        stage,
+        SyntheticStageOwner.STAGE_AFTER
+    )
+
+    // java .forEach rather than groovy .each, since the nested .each closure sometimes omits parent context
+    targetPercentages.forEach({ p ->
+      def resizeContext = baseContext + [
+          target: TargetServerGroup.Params.Target.current_asg_dynamic,
```

Comment:
Good question - my quick read of the target resolver stuff was that it would always resolve to whatever that first run of the `determineTargetServerGroup` stage found before all the resizes & disables start running, so even if something new was deployed it would keep resizing the same server group. But now that you mention it, the disable cluster stage will start disabling the wrong stuff... CheckPreconditions sounds like the best answer